### PR TITLE
(feat):Kyverno Adapter

### LINKED ISSUE #76: Admission Controller Adapter: Escape to Host
To prevent mounted docker socket: Kyverno has a policy - https://kyverno.io/policies/best-practices/disallow-cri-sock-mount/disallow-cri-sock-mount/

- Ensure containerized workloads do not run as root
- Create a Kyverno PSA policy in a restricted profile for namespaced scope.
- If possible create a ClusterScoped Policy which comes out of the box in Kyverno to apply it on multiple namespaces.

Ref:
https://github.com/DataDog/dirtypipe-container-breakout-poc

### PR DESCRIPTION
**Pull request by VedRatan**
## Description

 This PR introduces a new kyverno adapter for escapeToHost issue.

Fixes # (issue)
Closes #76 

**Does this PR introduce a breaking change?**

## Checklist

- [x] PR title follows the `<type>: <description>` convention
- [ ] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
- [ ] I have updated the [documentation](../docs) accordingly
- [ ] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
- [ ] I have performed a self-review of my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by anurag-rajawat** [comment]
In a follow-up PR add a helm chart and include this in our release adapters and adapter images build jobs:

https://github.com/5GSEC/nimbus/tree/main/deployments

https://github.com/5GSEC/nimbus/blob/b91563b0ccd36deee1c6afd73cf51fd2aa25aae8/.github/workflows/latest-release.yaml#L33-L34

https://github.com/5GSEC/nimbus/blob/b91563b0ccd36deee1c6afd73cf51fd2aa25aae8/.github/workflows/pr-checks.yaml#L85-L86

File: pkg/adapter/nimbus-kyverno/go.mod [hunk 1]
```diff
@@ -0,0 +1,283 @@
+module github.com/5GSEC/nimbus/pkg/adapter/nimbus-kyverno
+
+go 1.21.4
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/go.mod:3]: Please don't add patch version, keep it up to minor. e.g., `go 1.21`.

File: pkg/adapter/nimbus-kyverno/go.mod [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+toolchain go1.21.8
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/go.mod:5]: Remove this directive.

File: pkg/adapter/nimbus-kyverno/go.mod
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/go.mod:1]: 5gsec/nimbus dependencies are not added here. If you're using `go.work` file then please run `go work sync` from the nimbus root directory.

File: pkg/adapter/nimbus-kyverno/manager/manager.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+	}
+
+
+	deleteDanglingkcps(ctx, cnp, logger)
+	kcps := processor.BuildKcpsFrom(logger, &cnp)
+
+	for idx := range kcps {
+		kcp := kcps[idx]
+
+		// Set NimbusPolicy as the owner of the KSP
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/manager/manager.go:206]:
```suggestion
		// Set NimbusPolicy as the owner of the KCP
```

File: pkg/adapter/nimbus-kyverno/manager/manager.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+		delete(kpsToDelete, kpName)
+	}
+
+	for kpName := range kpsToDelete {
+		kp := kpsToDelete[kpName]
+		if err := k8sClient.Delete(ctx, &kp); err != nil {
+			logger.Error(err, "failed to delete dangling KyvernoPolicy", "KyvernoPolicy.Name", kp.Namespace, "KyvernoPolicy.Namespace", kp.Namespace)
+			continue
+		}
+
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/manager/manager.go:298]: After deleting a policy we have to update the corresponding nimbuspolicy status subresource to reflect the changes. Please call the API to update it.
For ref:
https://github.com/5GSEC/nimbus/blob/b91563b0ccd36deee1c6afd73cf51fd2aa25aae8/pkg/adapter/nimbus-netpol/manager/netpols_manager.go#L213-L215

File: pkg/adapter/nimbus-kyverno/manager/manager.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+		delete(kcpsToDelete, kcpName)
+	}
+
+	for kcpName := range kcpsToDelete {
+		kcp := kcpsToDelete[kcpName]
+		if err := k8sClient.Delete(ctx, &kcp); err != nil {
+			logger.Error(err, "failed to delete dangling KyvernoClusterPolicy", "KyvernoClusterPolicy.Name", kcp.Name)
+			continue
+		}
+
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/manager/manager.go:360]: Add a todo/reminder to remember to call the clusternimbuspolicy update status API in a follow-up PR to update the status subresource.

File: pkg/adapter/nimbus-kyverno/processor/kpbuilder.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	switch id {
+	case idpool.EscapeToHost:
+		return escapeToHost(np)
+	default:
+		return kyvernov1.Policy{}
+	}
+}
+
+
+
```
[the remark this answers is not in the document]
Comment (anurag-rajawat) [reply pkg/adapter/nimbus-kyverno/processor/kpbuilder.go:53]: Looks like you forgot to address this comment.

File: pkg/adapter/nimbus-kyverno/go.mod
Comment (VedRatan) [reply pkg/adapter/nimbus-kyverno/go.mod:1]: Its not reflecting the dependencies in the submodule

File: pkg/adapter/nimbus-kyverno/manager/manager.go [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+func createOrUpdateKcp(ctx context.Context, cnpName string) {
+	logger := log.FromContext(ctx)
+	var cnp intentv1.ClusterNimbusPolicy
+	if err := k8sClient.Get(ctx, types.NamespacedName{Name: cnpName}, &cnp); err != nil {
+		logger.Error(err, "failed to get ClusterNimbusPolicy", "ClusterNimbusPolicy.Name", cnpName)
+		return
+	}
+
+	if adapterutil.IsOrphan(cnp.GetOwnerReferences(), "ClusterSecurityIntentBinding") {
+		logger.V(4).Info("Ignoring orphan NimbusPolicy", "ClusterNimbusPolicy.Name", cnpName)
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/manager/manager.go:195]:
```suggestion
		logger.V(4).Info("Ignoring orphan ClusterNimbusPolicy", "ClusterNimbusPolicy.Name", cnpName)
```

File: pkg/adapter/nimbus-kyverno/manager/manager.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+	}
+
+
+	deleteDanglingkcps(ctx, cnp, logger)
+	kcps := processor.BuildKcpsFrom(logger, &cnp)
+
+	for idx := range kcps {
+		kcp := kcps[idx]
+
+		// Set NimbusPolicy as the owner of the KCP
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/manager/manager.go:206]:
```suggestion
		// Set ClusterNimbusPolicy as the owner of the KCP
```

File: pkg/adapter/nimbus-kyverno/processor/kpbuilder.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+func BuildKpsFrom(logger logr.Logger, np *v1.NimbusPolicy) []kyvernov1.Policy {
+	// Build KPs based on given IDs
+	var kps []kyvernov1.Policy
+	for _, nimbusRule := range np.Spec.NimbusRules {
+		id := nimbusRule.ID
+		if idpool.IsIdSupportedBy(id, "kyverno") {
+			kp := buildKpFor(id, np)
+			kp.Name = np.Name + "-" + strings.ToLower(id)+ "-" +strings.ToLower(id)
```
Comment (anurag-rajawat) [inline pkg/adapter/nimbus-kyverno/processor/kpbuilder.go:23]: Why are you appending `id` twice in kyverno policy name?

**Review by anurag-rajawat — COMMENTED** [review]
Rebase and some comments to address then we're good to merge this.

**Comment by anurag-rajawat** [comment]
Please address this [comment](https://github.com/5GSEC/nimbus/pull/77#issuecomment-2009148816).

**Comment by VedRatan** [comment]
@anurag-rajawat I had addressed your comments, I guess we are good to merge this one.

**Review by anurag-rajawat — APPROVED** [review]

**Pull request merged**